.gitmodules

```diff
@@ -31,3 +31,6 @@
 [submodule "lib/sdl/SDL2_image"]
 	path = lib/sdl/SDL2_image
 	url = https://github.com/libsdl-org/SDL_image.git
+[submodule "lib/usb/ohci_usbhostlib"]
+	path = lib/usb/ohci_usbhostlib
+	url = https://github.com/Ryzee119/ohci_usbhostlib.git
```

Comment:
An issue with the linux stack was that it wasn't easily possible to support new devices without upstreaming support, because the driver got built as part of nxdk. So if my app needed a new USB driver I always needed people to build a custom nxdk (example: Xbox 360 gamepads, but also potentially niche custom hardware for debugging which we might not want upstream).

So, can we register drivers out-of-tree somehow? (also why I brought up libusb in another comment)

I'm already not sure if mixing a generic HID driver + XID driver is good. I mean.. it works, but it just feels dirty and when more people want to add other devices it might explode in size.
If we want to share our code with others (which, ideally, we should, as mentioned in another comment), then it would help to have a scalable solution.

Reply:
Registering a new driver is very easy and can be done out of tree.
As simple as creating a `UDEV_DRV_T` struct
```
typedef struct
{
    int   (*probe) (struct iface_t *iface);  //what criteria should this driver be applied/init
    void  (*disconnect) (struct iface_t *iface);
    void  (*suspend) (struct iface_t *iface);
    void  (*resume) (struct iface_t *iface);
} UDEV_DRV_T;
```
`Then calling `usbh_register_driver(my_driver)`.

There's a bit of boilerplate code needed for each unique driver but not too bad.
HID and Xbox controller shared alot code, so just adding a simple class/subclass check in the `hid_probe` function seemed the easiest to me. This also has a benefit of just treating all HID devices in a similar way. I hope to  expose the steelbattalion via the joystick api. The xremote could be exposed in the same way.

Reply:
I still have to look into the other repository, but if a standard interface exists for adding USB drivers, then it might be worth it to add a similar interface for registering hid drivers (out-of-tree).

> I hope to expose the steelbattalion via the joystick api.

Are we talking about implementing the entire low-level driver in nxdk-sdl (SDL joystick), or are we talking about having a driver in the USB libs, with SDL using it?

I'd still prefer if the essential core drivers of nxdk were kept very small, so my idea of this would be:

* nxdk/lib/nxdk_usb.lib
* nxdk/lib/nxdk_usb_hid.lib
* nxdk/lib/nxdk_usb_hid_xid.lib
* nxdk/lib/nxdk_usb_hid_xid_steelbattalion.lib (if necessary)
* nxdk/lib/nxdk_sdl2.lib

Again: I'm probably overthinking it.